src/main.rs

```diff
@@ -36,7 +36,14 @@ pub mod expansion;
 pub struct Command {
     pub name: &'static str,
     pub help: &'static str,
-    pub main: Box<Fn(&Vec<String>, &mut BTreeMap<String, String>, &mut Vec<Mode>)>,
+    pub main: Box<Fn(&Vec<String>, &mut Shell)>,
```

Comment:
You might not want to take a mutable reference to `Shell` here. Just let the closure capture it.